src/line-top-index.js

```diff
@@ -69,53 +69,62 @@ export default class LineTopIndex {
     let oldEnd = traverse(start, oldExtent)
     let newEnd = traverse(start, newExtent)
 
+    let isInsertion = isZero(oldExtent)
     let startNode = this.iterator.insertNode(start)
-    let endNode = this.iterator.insertNode(oldEnd)
+    let endNode = this.iterator.insertNode(oldEnd, !isInsertion)
 
     startNode.priority = -1
     this.bubbleNodeUp(startNode)
     endNode.priority = -2
     this.bubbleNodeUp(endNode)
 
-    if (startNode !== endNode) {
-      let blockIdsToMove = new Set
+    startNode.blockIds.forEach((id) => {
+      if (!this.inclusiveBlockIds.has(id)) return
 
-      startNode.blockIds.forEach((id) => {
-        if (this.inclusiveBlockIds.has(id)) {
-          startNode.blockIds.delete(id)
-          blockIdsToMove.add(id)
-        }
-      })
+      startNode.blockIds.delete(id)
+      startNode.blockHeight -= this.blockHeightsById[id]
+      startNode.distanceFromLeftAncestor.pixels -= this.blockHeightsById[id]
 
-      if (startNode.right) {
-        this.collectBlockIdsForSubtree(startNode.right, blockIdsToMove)
-        startNode.right = null
-      }
+      endNode.blockIds.add(id)
+      endNode.blockHeight += this.blockHeightsById[id]
+      this.blockEndNodesById[id] = endNode
+    })
 
-      blockIdsToMove.forEach(id => {
+    if (startNode.right) {
+      let blockIdsToMove = new Set
+      this.collectBlockIdsForSubtree(startNode.right, blockIdsToMove)
+      blockIdsToMove.forEach((id) => {
         endNode.blockIds.add(id)
         endNode.blockHeight += this.blockHeightsById[id]
         this.blockEndNodesById[id] = endNode
       })
+      startNode.right = null
     }
 
     endNode.distanceFromLeftAncestor.point = newEnd
 
+    if (comparePoints(startNode.distanceFromLeftAncestor.point, endNode.distanceFromLeftAncestor.point) === 0) {
```

Comment:
Yeah, basically when there isn't a new extent we want to merge the two nodes . I took inspiration from the marker-index but we could also extract a variable to make this explicit.